### PR TITLE
[Feature] jwts for waterbutler

### PR DESCRIPTION
**Pull request by chrisseto**
Use jwts for waterbutler credentials endpoint and remove restrict_waterbutler

OSF-4733

### CODE REVIEW

File: tests/test_addons.py [hunk 1]
```diff
@@ -128,53 +126,45 @@ def configure_addon(self):
         self.node_addon.save()
 
     def build_url(self, **kwargs):
-        options = dict(
+        options = {'payload': jwt.encode(dict(dict(
```
Comment (icereval) [inline tests/test_addons.py]: For added security set expiration to 30s (https://github.com/jpadilla/pyjwt#expiration-time-claim)

File: website/addons/base/views.py [hunk 2]
```diff
@@ -202,20 +191,20 @@ def get_auth(auth, **kwargs):
...
         'auth': make_auth(auth.user),  # A waterbutler auth dict not an Auth object
         'credentials': credentials,
-        'settings': settings,
+        'settings': waterbutler_settings,
         'callback_url': node.api_url_for(
             ('create_waterbutler_log' if not node.is_registration else 'registration_callbacks'),
             _absolute=True,
         ),
-    }
+    }, settings.WATERBUTLER_JWT_SECRET, algorithm=settings.WATERBUTLER_JWT_ALGORITHM)
```
Comment (icereval) [inline website/addons/base/views.py]: For added security set expiration to 30s (https://github.com/jpadilla/pyjwt#expiration-time-claim)

File: tests/test_addons.py (hunk 1 above)
Comment (icereval) [inline tests/test_addons.py]: Should use `jwt.encode` on the entire json object.

File: tests/test_addons.py [hunk 3]
```diff
@@ -128,53 +126,45 @@ def configure_addon(self):
...
             action='download',
-            cookie=self.cookie,
             nid=self.node._id,
             provider=self.node_addon.config.short_name,
-        )
-        options.update(kwargs)
+        ), **kwargs), settings.WATERBUTLER_JWT_SECRET, algorithm=settings.WATERBUTLER_JWT_ALGORITHM)}
```
Comment (icereval) [inline tests/test_addons.py]: For added security set expiration to 30s (https://github.com/jpadilla/pyjwt#expiration-time-claim)

File: website/addons/base/views.py [hunk 4]
```diff
@@ -184,9 +168,14 @@ def get_auth(auth, **kwargs):
             auth.user = User.from_cookie(request.args.get('cookie'))
 
     try:
-        action = request.args['action']
-        node_id = request.args['nid']
-        provider_name = request.args['provider']
+        data = jwt.decode(request.args.get('payload', ''), settings.WATERBUTLER_JWT_SECRET, algorithm=settings.WATERBUTLER_JWT_ALGORITHM)
```
Comment (icereval) [inline website/addons/base/views.py]: same as `jwt.encode`, entire content should be the encrypted JWT.

File: website/addons/base/views.py [hunk 5]
```diff
@@ -184,9 +169,14 @@ def get_auth(auth, **kwargs):
...
```
Comment (icereval) [inline website/addons/base/views.py]: Require exp - https://github.com/jpadilla/pyjwt#requiring-optional-claims

File: website/addons/base/views.py [hunk 6]
```diff
@@ -202,20 +192,21 @@ def get_auth(auth, **kwargs):
...
         credentials = provider_settings.serialize_waterbutler_credentials()
-        settings = provider_settings.serialize_waterbutler_settings()
+        waterbutler_settings = provider_settings.serialize_waterbutler_settings()
     except exceptions.AddonError:
         log_exception()
         raise HTTPError(httplib.BAD_REQUEST)
 
-    return {
+    return jwt.encode({
+        'exp': datetime.datetime.utcnow() + datetime.timedelta(seconds=settings.WATERBUTLER_JWT_EXPIRATION),
```
Comment (icereval) [inline website/addons/base/views.py]: This is very unlikely but I'll at the least post the question. Any concern for key collision, should we nest our data under `data`? https://github.com/jpadilla/pyjwt#support-of-registered-claim-names

**Comment by icereval** [comment]
First review done :pill:

File: website/settings/defaults.py [hunk 7]
```diff
@@ -269,3 +269,7 @@ def parent_dir(path):
 
 JWT_SECRET = 'changeme'
 JWT_ALGORITHM = 'HS256'
+
+WATERBUTLER_JWT_SECRET = 'ILiekTrianglesALot'
+WATERBUTLER_JWT_ALGORITHM = 'HS256'
+WATERBUTLER_JWT_EXPIRATION = 15
```
Comment (icereval) [inline website/settings/defaults.py]: WATERBUTLER_ADDRS = ['127.0.0.1']  can be removed

**Comment by sloria** [comment]
2nd pass finished, Good when Michael's comments are addressed.

File: website/settings/__init__.py [hunk 8]
```diff
@@ -6,10 +6,16 @@
...
-    from .local import *
+    from .local import *  # noqa
 except ImportError as error:
     raise ImportError("No local.py settings file found. Did you remember to "
                         "copy local-dist.py to local.py?")
+
+if not DEV_MODE:
+    from . import local
+    from . import defaults
+    for setting in ('WATERBUTLER_JWT_SECRET', 'JWT_SECRET', 'DB_PASS', 'DEFAULT_HMAC_SECRET'):
```
Comment (sloria) [inline website/settings/__init__.py]: :+1:

**Pull request merged**